pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -90,117 +104,283 @@ func (c *ClusterMemberController) reconcileMembers(ctx context.Context, recorder
 	}
+
 	return nil
 }
 
-func (c *ClusterMemberController) isMachinePendingDeletionFor(etcdHost string) (bool, error) {
-	if isFunctional, err := c.machineAPIChecker.IsFunctional(); err != nil {
-		return false, err
-	} else if !isFunctional {
-		// always return false when the machine API is off
-		// otherwise we won't be able to add any member
-		return false, nil
+// getEtcdPeerURLToAdd returns a PeerURL of the next etcd instance to be added as a learner to the cluster. If
+// no eligible member is found return an empty string.
+func (c *ClusterMemberController) getEtcdPeerURLToAdd(ctx context.Context) (string, error) {
+	members, err := c.etcdClient.MemberList(ctx)
+	if err != nil {
+		return "", err
 	}
 
-	masterMachines, err := c.masterMachineLister.List(c.masterMachineSelector)
+	nodes, err := c.masterNodeLister.List(c.masterNodeSelector)
 	if err != nil {
-		return false, err
+		return "", err
 	}
 
-	// one we have functional machine API, get the corresponding machine and check DeletionTimestamp
-	switch machineForEtcdHost, hasMachine := ceohelpers.IndexMachinesByNodeInternalIP(masterMachines)[etcdHost]; {
-	case hasMachine && machineForEtcdHost.DeletionTimestamp != nil:
-		klog.V(2).Infof("member: %v has a machine that is pending deletion: %v", etcdHost, machineForEtcdHost.Name)
-		return true, nil
-	case !hasMachine:
-		return false, fmt.Errorf("unable to find machine for member: %v", etcdHost)
-	default:
-		// we've found a machine and it is not pending deletion
-		return false, nil
+	for _, node := range nodes {
+		if node.DeletionTimestamp != nil {
+			klog.V(2).Infof("Ignoring node (%s) for new member addition as it's pending deletion", node.Name)
+			continue
+		}
+
+		peerURL, err := c.getPeerURLForNode(node)
+		if err != nil {
+			return "", fmt.Errorf("failed to get peerURL for node: %v", err)
+		}
+
+		// Ignore if the node's PeerURL is mapped to a member and already part of the quorum.
+		if isURLMappedToMember(peerURL, members) {
+			continue
+		}
+
+		// If the Machine API is unavailable then we can't evaluate the machine's
+		// deletion timestamp or deletion hooks so we scale it up without any considerations
+		// for vertical scaling
+		// This allows UPI clusters without a machine API to scale up
+		if !c.isMachineAPIFunctional {
+			return peerURL, nil
+		}
+
+		// If the Machine API is functional we should ignore the machine
+		// if it is pending deletion to avoid re-adding previously removed members
+		internalIP, err := c.getInternalIPForNode(node)
+		if err != nil {
+			return "", fmt.Errorf("failed to get internal IP for node: %v", err)
+		}
+		machine, err := c.findMachineByNodeInternalIP(internalIP)
+		if err != nil {
+			return "", fmt.Errorf("failed to get machine for node (%s): %v", node.Name, err)
+		}
+		if machine == nil {
+			klog.V(2).Infof("Ignoring node (%s) for scale-up: no Machine found referencing this node's internal IP (%v)", node.Name, internalIP)
+			continue
+		}
```

Comment:
@p0lyn0mial Just a side note: I've had to had to refactor `isMachinePendingDeletionFor()` from #790 into the conditionals here. We're no longer erroring out if we don't find a machine, since we want to move on to attempt with adding other eligible nodes as learners.
Having a log statement is good enough imo, and if the machine appears, we'll retry anyway.

Also updated the associated unit test case for this from #790